File: src/pretix/helpers/jsonlogic_boolalg.py
#
# This file is part of pretix (Community Edition).
#
# Copyright (C) 2014-2020 Raphael Michel and contributors
# Copyright (C) 2020-2021 rami.io GmbH and contributors
#
# This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General
# Public License as published by the Free Software Foundation in version 3 of the License.
#
# ADDITIONAL TERMS APPLY: Pursuant to Section 7 of the GNU Affero General Public License, additional terms are
# applicable granting you additional permissions and placing additional restrictions on your usage of this software.
# Please refer to the pretix LICENSE file to obtain the full terms applicable to this work. If you did not receive
# this file, see <https://pretix.eu/about/en/license>.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along with this program.  If not, see
# <https://www.gnu.org/licenses/>.
#
import logging

logger = logging.getLogger(__name__)


def convert_to_dnf(rules):
    """
    Converts a set of rules to disjunctive normal form, i.e. returns something of the form
    `(a AND b AND c) OR (a AND d AND f)`
    without further nesting.
    """
    if not isinstance(rules, dict):
        return rules

    def _distribute_or_over_and(r):
        operator = list(r.keys())[0]
        values = r[operator]
        if operator == "and":
            arg_to_distribute = [arg for arg in values if isinstance(arg, dict) and "or" in arg]
            if not arg_to_distribute:
                return rules
            arg_to_distribute = arg_to_distribute[0]
            other_args = [arg for arg in values if arg is not arg_to_distribute]
            return {
                "or": [
                    {"and": [*other_args, dval]} for dval in arg_to_distribute["or"]
                ]
            }
        elif operator in ("!", "!!", "?:", "if"):
            raise ValueError(f"Operator {operator} currently unsupported by convert_to_dnf")
        else:
            return r

    def _simplify_chained_operators(r):
        # Simplify `(a OR b) OR (c or d)` to `a OR b OR c OR d` and the same with `AND`
        if not isinstance(r, dict):
            return r
        operator = list(r.keys())[0]
        values = r[operator]
        if operator not in ("or", "and"):
            return r
        new_values = []
        for v in values:
            if not isinstance(v, dict) or operator not in v:
                new_values.append(v)
            else:
                new_values += v[operator]
        return {operator: new_values}

    # Run _distribute_or_over_and on until it no longer changes anything. Do so recursively
    # for the full expression tree.
    old_rules = rules
    while True:
        rules = _distribute_or_over_and(rules)
        operator = list(rules.keys())[0]
        values = rules[operator]
        no_list = False
        if not isinstance(values, list):
            values = [values]
            no_list = True
        rules = {
            operator: [
                convert_to_dnf(v) for v in values
            ] if not no_list else convert_to_dnf(values[0])
        }
        if old_rules == rules:
            break
        old_rules = rules
    # Simplify leftovers of the recursion
    rules = _simplify_chained_operators(rules)
    return rules

File: src/pretix/static/pretixcontrol/js/ui/checkinrules/jsonlogic-boolalg.js
export function convertToDNF (rules) {
	// Converts a set of rules to disjunctive normal form, i.e. returns something of the form
	// `(a AND b AND c) OR (a AND d AND f)`
	// without further nesting.
	if (typeof rules !== 'object' || Array.isArray(rules) || rules === null) {
		return rules
	}

	function _distribute_or_over_and (r) {
		let operator = Object.keys(r)[0]
		let values = r[operator]
		if (operator === 'and') {
			let arg_to_distribute = null
			let other_args = []
			for (let arg of values) {
				if (typeof arg === 'object' && !Array.isArray(arg) && typeof arg['or'] !== 'undefined' && arg_to_distribute === null) {
					arg_to_distribute = arg
				} else {
					other_args.push(arg)
				}
			}
			if (arg_to_distribute === null) {
				return r
			}
			let or_operands = []
			for (let dval of arg_to_distribute['or']) {
				or_operands.push({ and: other_args.concat([dval]) })
			}
			return {
				or: or_operands
			}
		} else if (!operator) {
			return r
		} else if (operator === '!' || operator === '!!' || operator === '?:' || operator === 'if') {
			console.warn('Operator ' + operator + ' currently unsupported by convert_to_dnf')
			return r
		} else {
			return r
		}
	}

	function _simplify_chained_operators (r) {
		// Simplify `(a OR b) OR (c or d)` to `a OR b OR c OR d` and the same with `AND`
		if (typeof r !== 'object' || Array.isArray(r)) {
			return r
		}
		let operator = Object.keys(r)[0]
		let values = r[operator]
		if (operator !== 'or' && operator !== 'and') {
			return r
		}
		let new_values = []
		for (let v of values) {
			if (typeof v !== 'object' || Array.isArray(v) || typeof v[operator] === 'undefined') {
				new_values.push(v)
			} else {
				new_values.push(...v[operator])
			}
		}
		let result = {}
		result[operator] = new_values
		return result
	}

	// Run _distribute_or_over_and on until it no longer changes anything. Do so recursively
	// for the full expression tree.
	let old_rules = rules
	while (true) {
		rules = _distribute_or_over_and(rules)
		let operator = Object.keys(rules)[0]
		let values = rules[operator]
		let no_list = false
		if (!Array.isArray(values)) {
			values = [values]
			no_list = true
		}
		rules = {}
		if (!no_list) {
			rules[operator] = []
			for (let v of values) {
				rules[operator].push(convertToDNF(v))
			}
		} else {
			rules[operator] = convertToDNF(values[0])
		}
		if (JSON.stringify(old_rules) === JSON.stringify(rules)) { // Let's hope this is good enough...
			break
		}
		old_rules = rules
	}
	rules = _simplify_chained_operators(rules)
	return rules
}
